sql/mysqld.cc

```diff
@@ -1730,6 +1730,12 @@ static void close_connections(void)
   while (CONNECT::count)
     my_sleep(100);
 
+  /*
+    If the signal thread is actively using server resources, let it finish
+    before killing any connections
+  */
+  wait_for_signal_thread_to_end();
```

Comment:
I think it makes more sense to put that one at the very end of the function. Advantages are:
- it shuts down use/background THDs earlier, rather than waiting in worst case for 10 sec before informing them about shutdown.
- there is less CPU load at the end of this function (all THDs are shut down), so chances are high 10 seconds will be enough for  `wait_for_signal_thread_to_end()`  to finish, even if it had to flushing table cache.